### DISCUSSION TITLE
Thanks a lot for your work!

### DISCUSSION THREAD

**Discussion by xluckydegen** [CoderLine/mocha-vscode#31] [General]
Hello,

I want to thank you very much for your work on this extension. Till now I used outdated mocha explorer, which doesn't work with node 20+.

Although I found few minor glitches in your plugin, I have to say it's very pleasant to work with it in comparison to mentioned explorer, where configuration was always pain.

My observations during first few hours of configuration and migration to the project:

- configuration based on mocharc is very user friendly
- output logs are perfect for testing, including visible full executable path which helps with a debugging.

Few glitches I found:

- when .mocharc.js is changed to .mocharc.cjs during runtime, plugin still searching for .js during test (altghout refresh of tests in test explorer works fine, executing tests will fail until vscode is restarted)
- when using example mocha file from here (https://github.com/mochajs/mocha/blob/master/example/config/.mocharc.js) plugin didn't executed tests.  Then I started with example from this issue (https://github.com/CoderLine/mocha-vscode/issues/23) and everything started working

Few ideas how to improve it:

- it would be nice to have separated console output like original mocha explorer offered. Meaning that you can click on test and display isolated output only for one test.

And that's all for now. I will post more observations as I will continue with migration process.

Thanks

**Comment by xluckydegen** [comment]
Hello, I have one new observation. I have following demo tests:

```

import { expect } from "chai";

describe("Demo TS", function ()
{
  it("put new", async function () 
  {
    console.log("logggg");
    expect(true).eq(true)
  });

  it("put new 2", async function () 
  {
    console.log("logggg22222");
    expect(true).eq(true)
  });
});
```

When I start tests, in Test Results window I see console.log outputs like this:

![Code_2024-04-12_08-15-52](https://github.com/CoderLine/mocha-vscode/assets/116631887/a6c7ded7-1309-44de-9891-cd2520489fa1)

But when I click on result tree, where I expected that I will see the console.log for 1 test only, the result is:  `The test case did not report any output. `

![ShareX_2024-04-12_08-16-59](https://github.com/CoderLine/mocha-vscode/assets/116631887/d9d1b335-ff53-4cd1-be67-18ddb2441997)

Is this expected behavior?  Is there any way how to write console.log to this test-result also per-test? Thanks

**Reply by muhammadyusuf-kurbonov** [reply]
https://github.com/CoderLine/mocha-vscode/pull/137

**Comment by xluckydegen** [comment]
Few more:

- Show skipped tests/suits with different icon in tree (not only after test are executed but directly in tree after loading it).
- When a whole test-suit is skipped (root describe.skip is used), there is error in the test runner:
- The same when all tests inside the file are commented and only root descrption() is available

![Code_2024-04-12_08-22-06](https://github.com/CoderLine/mocha-vscode/assets/116631887/9a298ffa-f845-4aba-a9b9-a2f1e8d9cf98)

![Code_2024-04-12_08-22-28](https://github.com/CoderLine/mocha-vscode/assets/116631887/7933fe0f-e587-4da8-a89f-78bd138f081b)

**Comment by xluckydegen** [comment]
Few more issues:

Running more complex test on multiple files lead to issue with mocha:

![Code_2024-04-12_12-50-54](https://github.com/CoderLine/mocha-vscode/assets/116631887/09fe96c7-82e8-4c47-933f-34ab8ef8e7a2)
![Code_2024-04-12_12-52-02](https://github.com/CoderLine/mocha-vscode/assets/116631887/aa46c086-572b-4aaa-a979-da9b1608d2fb)

the same issue occurs when I run generated mocha command manually from command line. It seems that mocha doesn't support passing multiple --file commands

**Reply by xluckydegen** [reply]
I just tried to downgrade to node 18.18 and mocha is now showing more info. Not sure if it's related to the issue or different one, but now I see this when selecting multiple files (when executed per single file, everything still works ok):

```
 >  C:\Program Files\nodejs\node.EXE q:\ApplicationsNodeJs\App\node_modules\mocha\bin\mocha.js --config q:\ApplicationsNodeJs\App\.mocharc.cjs --reporter q:\ApplicationsNodeJs\App\node_modules\.mocha-vscode\fullJsonStreamReporter.js --ignore **/*.* --file shared\appservices\servicemerkletree\logic\logicservicemerkletree.test.ts --file shared\appservices\servicemerkletree\logic\merkletree\merkletree.test.ts}

TSError: ⨯ Unable to compile TypeScript:
shared/appservices/servicemerkletree/logic/merkletree/merkletree.test.ts(3,33): error TS1149: File name 'q:/ApplicationsNodeJs/App/shared/appservices/servicemerkletree/logic/merkletree/merkleTreeCache.ts' differs from already included file name 'q:/ApplicationsNodeJs/App/shared/appservices/servicemerkletree/logic/merkleTree/merkleTreeCache.ts' only in casing.
  The file is in the program because:
    Imported via "./merkleTree/merkleTreeCache" from file 'q:/ApplicationsNodeJs/App/shared/appservices/servicemerkletree/logic/logicServiceMerkleTree.ts'
    Imported via "./merkleTreeCache" from file 'q:/ApplicationsNodeJs/App/shared/appservices/servicemerkletree/logic/merkletree/merkletree.test.ts'

    at createTSError (q:\ApplicationsNodeJs\App\node_modules\ts-node\src\index.ts:859:12)
    at reportTSError (q:\ApplicationsNodeJs\App\node_modules\ts-node\src\index.ts:863:19)
    at getOutput (q:\ApplicationsNodeJs\App\node_modules\ts-node\src\index.ts:1077:36)
    at Object.compile (q:\ApplicationsNodeJs\App\node_modules\ts-node\src\index.ts:1433:41)
    at transformSource (q:\ApplicationsNodeJs\App\node_modules\ts-node\src\esm.ts:400:37)
    at q:\ApplicationsNodeJs\App\node_modules\ts-node\src\esm.ts:278:53
    at async addShortCircuitFlag (q:\ApplicationsNodeJs\App\node_modules\ts-node\src\esm.ts:409:15)
    at async nextLoad (node:internal/modules/esm/loader:163:22)
    at async ESMLoader.load (node:internal/modules/esm/loader:603:20)
    at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:457:11)
    at async link (node:internal/modules/esm/module_job:68:21)
Error: Test process exited with code 1
```

**Comment by Danielku15** [comment]
Hey there 👋 

Nice to see already some adoption of the plugin. Here some feedback to your posts:

> when .mocharc.js is changed to .mocharc.cjs during runtime, plugin still searching for .js during test (altghout refresh of tests in test explorer works fine, executing tests will fail until vscode is restarted)

I added https://github.com/CoderLine/mocha-vscode/issues/32 to track this. Will investigate and fix it soon. 

> when using example mocha file from here (https://github.com/mochajs/mocha/blob/master/example/config/.mocharc.js) plugin didn't executed tests. Then I started with example from this issue (https://github.com/CoderLine/mocha-vscode/issues/23) and everything started working

Can you setup an example repository and double check this problem? If its reproducible like that would be great to enter a bug via https://github.com/CoderLine/mocha-vscode/issues/new/choose with all the details. 

> it would be nice to have separated console output like original mocha explorer offered. Meaning that you can click on test and display isolated output only for one test.

VS Code has an own built-in mechanism for that. We attach any captured outputs during the test execution to the individual tests (like the outcome) and the output can be checked when selecting the test in the bottom panel. Its a bit different to the "older" plugin but how Microsoft has designed it. If something is not working fully correct there let me know, could be that there are some bugs in capturing the output. 

For any bugs you encounter please enter them via https://github.com/CoderLine/mocha-vscode/issues/new/choose ot have all information collected and trackable. 

> Show skipped tests/suits with different icon in tree (not only after test are executed but directly in tree after loading it).

Skipped tests are indeed a bit tricky. Unfortunately Mocha doesn't report them at all in the execution so that might lead to errors. I added https://github.com/CoderLine/mocha-vscode/issues/33

> Running more complex test on multiple files lead to issue with mocha:

Can you file a bug for that with an example repo? We already have quite some tests for checking the test combinations and most variants are fine. This might be a missing case I have to narrow down precisely. 

https://github.com/CoderLine/mocha-vscode/blob/main/src/test/integration/overlapping-tests.test.ts

**Comment by xluckydegen** [comment]
Hello, thank you for detailed reply.

I will try to create sample repositories later, but I'm not sure I will be able to find the correct combination. I tried a lot of things to make all things work and didn't save the non-working ones. Also I still testing it on our production repo, so I have to create some complete new repo. But I will try it to help you.

Regarding bugs, I will create them there. Originally I didn't wanted to spam the bug tracker with issues in case you already know about that or when it should be caused by some of my own wrong configs.

Regarding skipping tests, I'm not sure how, but original Mocha Test explorer is able to do that even on very large projects, Unfortunately I'm originally c++ dev and NodeJS is still pretty new for me, so I'm still little lost in these advanced configurations.

Thanks

**Reply by Danielku15** [reply]
I'd say at this point of development: any bug is likely a valid bug I'm not fully aware of. There are some I had encountered but still its better to enter them all if you can reproduce them.

Fine if you spam the bug tracker. The 'worst' which might happen is that I add a unit test to proof that things are working and close the bug 😁
